server/pulp/plugins/file/model_distributor.py

```diff
@@ -0,0 +1,249 @@
+                for unit in units:
+                    links_to_create = self.get_paths_for_unit(unit)
+                    self._symlink_unit(build_dir, unit, links_to_create)
+                    self.publish_metadata_for_unit(unit)
+            finally:
+                # Finalize the processing
+                self.finalize_metadata()
+
+            # Let's unpublish, and then republish
+            self.unpublish_repo(repo_model, config)
+
+            hosting_locations = self.get_hosting_locations(repo_model, config)
+            for location in hosting_locations:
+                shutil.copytree(build_dir, location, symlinks=True)
+
+            self.post_repo_publish(repo_model, config)
+
+            # Clean up our build_dir
+            common_utils.delete_working_directory()
+
+            # Report that we are done
+            progress_report.state = progress_report.STATE_COMPLETE
+            return progress_report.build_final_report()
+        except Exception, e:
+            _logger.exception(e)
+            # Something failed. Let's put an error message on the report
+            progress_report.error_message = str(e)
+            progress_report.traceback = traceback.format_exc()
+            progress_report.state = progress_report.STATE_FAILED
+            report = progress_report.build_final_report()
+            return report
+
+    def unpublish_repo(self, repo, config):
+        """
+        Delete the published files from our filesystem
+
+        Please also see the superclass method definition for more documentation on this method.
+
+        :param repo: Repository model object
+        :type  repo: pulp.server.db.model.Repository
+
+        :param config: plugin configuration
+        :type  config: pulp.plugins.config.PluginCallConfiguration
+        """
+        hosting_locations = self.get_hosting_locations(repo, config)
+        for location in hosting_locations:
+            self._rmtree_if_exists(location)
+
+    def validate_config(self, repo, config, config_conduit):
+        raise NotImplementedError()
+
+    def initialize_metadata(self, build_dir):
+        """
+        Initialize the metadata files that are going to be used for
+        :param build_dir: The working directory where the new metadata should be generated
+        :type build_dir: str
+        """
+        metadata_filename = os.path.join(build_dir, MANIFEST_FILENAME)
+        self.metadata_file = open(metadata_filename, 'w')
+        self.metadata_csv_writer = csv.writer(self.metadata_file)
+
+    def publish_metadata_for_unit(self, unit):
+        """
+        Publish the metadata for a single unit.
+        This should be writing to open file handles from the initialize_metadata call
+
+        :param unit: the unit for which metadata needs to be generated
+        :type unit: pulp.server.db.model.ContentUnit
+        """
+        self.metadata_csv_writer.writerow([unit.unit_key['name'], unit.unit_key['checksum'],
```

Comment:
This way will work, but I usually use `unit.name` instead of `unit.unit_key['name']`. Similarly for `checksum` and `size`.